src/Locator/RootsFileOrDirectoryLocator.php

```diff
@@ -33,66 +33,89 @@
 
 declare(strict_types=1);
 
-namespace Infection\Finder;
+namespace Infection\Locator;
 
-use Infection\Finder\Exception\LocatorException;
+use function current;
+use const DIRECTORY_SEPARATOR;
 use function Safe\realpath;
 use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\Assert\Assert;
+use Webmozart\PathUtil\Path;
 
 /**
  * @internal
  */
-final class Locator implements LocatorInterface
+final class RootsFileOrDirectoryLocator implements Locator
 {
-    /**
-     * @var string[]
-     */
-    private $paths;
+    private $roots;
+    private $filesystem;
 
     /**
-     * @var Filesystem
+     * @param  string[] $roots
      */
-    private $filesystem;
-
-    public function __construct(array $paths, Filesystem $filesystem)
+    public function __construct(array $roots, Filesystem $filesystem)
     {
-        $this->paths = $paths;
+        Assert::allString($roots);
+
+        $this->roots = $roots;
         $this->filesystem = $filesystem;
     }
 
-    public function locate(string $name): string
+    /**
+     * {@inheritdoc}
+     */
+    public function locate(string $fileName): string
     {
-        if ($this->filesystem->isAbsolutePath($name)) {
-            if ($this->filesystem->exists($name)) {
-                return realpath($name);
+        $canonicalFileName = Path::canonicalize($fileName);
+
+        if ($this->filesystem->isAbsolutePath($canonicalFileName)) {
+            if ($this->filesystem->exists($canonicalFileName)) {
+                return realpath($canonicalFileName);
             }
 
-            throw LocatorException::fileOrDirectoryDoesNotExist($name);
+            throw FileNotFound::createForFile($canonicalFileName, $this->roots);
         }
 
-        foreach ($this->paths as $path) {
-            $file = $path . \DIRECTORY_SEPARATOR . $name;
+        foreach ($this->roots as $path) {
+            $file = $path . DIRECTORY_SEPARATOR . $canonicalFileName;
 
             if ($this->filesystem->exists($file)) {
                 return realpath($file);
             }
         }
 
-        throw LocatorException::filesOrDirectoriesDoNotExist($name, $this->paths);
+        throw FileNotFound::createForFile($canonicalFileName, $this->roots);
     }
 
+    /**
+     * {@inheritdoc}
+     */
     public function locateOneOf(array $fileNames): string
     {
-        if (!$fileNames) {
-            throw LocatorException::filesNotFound();
+        $file = $this->innerLocateOneOf($fileNames);
+
+        if (null === $file) {
+            throw FileNotFound::createForFiles($fileNames, $this->roots);
+        }
+
+        return $file;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    private function innerLocateOneOf(array $fileNames): ?string
+    {
+        if ([] === $fileNames) {
+            return null;
```

Comment:
Perhaps an empty `string` would be a better option than using `null`?

Reply:
I personally find using an empty string over `null` more error prone, but I don't feel strongly about it either so if others feel that way too I'm fine with changing it

Reply:
At least we could get rid of the nullable return type declaration.

Reply:
It is on purpose though.

In the public method `locateOneOf()`, it is "get the result or throw" whereas here it's the internal version which is "try to _find_ a result", i.e. not having a result is a happy path (consequence of the recursion)